web/pandas/pdeps/0007-copy-on-write.md

```diff
@@ -0,0 +1,575 @@
+`df2` or `df3` would mutate `df`. Making use of Copy-on-Write allows us to also avoid
+unnecessary copies in methods such as the above (or in the variant using method chaining
+like `df.rename(columns=str.lower).set_index("a")`).
+
+### Propagating mutation forwards
+
+Thus far we have considered the (more common) case of taking a subset, mutating the
+subset, and how that should affect the parent. What about the other direction, where the
+parent is mutated?
+
+```python
+>>> df = pd.DataFrame({"A": [1, 2], "B": [3, 4]})
+>>> df2 = df[["A"]]
+>>> df.iloc[0, 0] = 10
+>>> df2.iloc[0, 0]  # what is this value?
+```
+
+Given that `df2` is _considered_ as a copy of df under this proposal (i.e. behaves as a
+copy), also mutating the parent `df` will not mutate the subset `df2`.
+
+### When do mutations propagate to other objects and when not?
+
+This proposal basically means that mutations _never_ propagate to _other_ objects (as
+would happen with views). The only way to modify a DataFrame or Series is to modify the
+object itself directly.
+
+But let's illustrate this in Python terms. Consider that we have a DataFrame df1, and we
+assign that to another name df2:
+
+```python
+>>> df1 = pd.DataFrame({"A": [1, 2], "B": [3, 4]})
+>>> df2 = df1
+```
+
+Although we have now two variables (df1 and df2), this assignment follows the standard
+python semantics, and both names are pointing to the same object ("df1 and df2 are
+_identical_"):
+
+```python
+>>> id(df1) == id(df2)  # or: df1 is df2
+True
+```
+
+Thus, if you modify DataFrame df2, this is also reflected in the other variable df1, and
+the other way around (since it's the same object):
+
+```python
+>>> df1.iloc[0, 0]
+1
+>>> df2.iloc[0, 0] = 10
+>>> df1.iloc[0, 0]
+10
+```
+
+In summary, modifications are only "propagated" between _identical_ objects (not just
+equal (`==`), but identical (`is`) in python terms
+([docs](https://docs.python.org/3/reference/expressions.html#is))). Propagation is not
+really the proper term, since there is only one object that was modified.
+
+However, when in some way creating a new object (even though it might be a DataFrame
+with the same data, and thus be an "equal" DataFrame):
+
+```python
+>>> df1 = pd.DataFrame({"A": [1, 2], "B": [3, 4]})
+>>> df2 = df1[:]  # or df1.loc[...] with some indexer
+```
+
+Those objects are no longer identical:
+
+```python
+>>> id(df1) == id(df2)  # or df1 is df2
+False
+```
+
+And thus modifications to one will not propagate to the other:
+
+```python
+>>> df1.iloc[0, 0]
+1
+>>> df2.iloc[0, 0] = 10
+>>> df1.iloc[0, 0]  # not changed
+1
+```
+
+In pandas, any getitem indexing operation returns _new_ objects, and also almost all
+DataFrame/Series methods return a _new_ object (except with `inplace=True` in some
+cases), and thus follow the above logic of never modifying its parent/child DataFrame or
+Series (using the lazy Copy-on-Write mechanism where possible).
```

Comment:
For the above paragraph, it isn't clear when you say "In pandas, ..." you are talking about current behavior or the proposed behiavior.

Reply:
Both, every operation returns a new object, with CoW this means never updating two objects at once. Right now this has no particular effect

Reply:
Nevertheless, I think this should be clarified in the text.